### PR TITLE
Add (aws) s3 as media backend for Mastodon

### PR DESCRIPTION
**Pull request by Yornik**
###### Description of changes

Added the options to run Matsadon with a s3 backend

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: nixos/modules/services/web-apps/mastodon.nix [hunk 1]
```diff
@@ -494,6 +538,9 @@ in {
         VAPID_PRIVATE_KEY="$(cat ${cfg.vapidPrivateKeyFile})"
         VAPID_PUBLIC_KEY="$(cat ${cfg.vapidPublicKeyFile})"
         DB_PASS="$(cat ${cfg.database.passwordFile})"
+      '' + (if cfg.aws.accessKeyID then ''
+        AWS_SECRET_ACCESS_KEY="$(cat ${cfg.aws.SecretAccessKeyFile})"
+      '' else "") + ''
       '' + (if cfg.smtp.authenticate then ''
         SMTP_PASSWORD="$(cat ${cfg.smtp.passwordFile})"
       '' else "") + ''
```
Comment (SuperSandro2000) [inline nixos/modules/services/web-apps/mastodon.nix:546]: This should use lib.optionalString

File: nixos/modules/services/web-apps/mastodon.nix [hunk 2]
```diff
@@ -403,6 +409,44 @@ in {
...
+          example = "/run/keys/mastodon-smtp-password";
+          type = lib.types.str;
+        };
+        accessKeyID = lib.mkOption {
+          description = lib.mdDoc "aws access key ID.";
+          type = lib.types.str;
+        };
+      };
+
+      s3 = {
```
Comment (SuperSandro2000) [inline nixos/modules/services/web-apps/mastodon.nix:427]: Those are missing descriptions

File: nixos/modules/services/web-apps/mastodon.nix [hunk 3]
```diff
@@ -20,20 +20,26 @@ let
...
+    DB_PORT = toString(cfg.database.port);
     ES_ENABLED = if (cfg.elasticsearch.host != null) then "true" else "false";
     ES_HOST = cfg.elasticsearch.host;
     ES_PORT = toString(cfg.elasticsearch.port);
+    LOCAL_DOMAIN = cfg.localDomain;
+    PAPERCLIP_ROOT_PATH = "/var/lib/mastodon/public-system";
+    PAPERCLIP_ROOT_URL = "/system";
+    REDIS_HOST = cfg.redis.host;
+    REDIS_PORT = toString(cfg.redis.port);
+    S3_ENABLED= if (cfg.s3.bucket != null) then "true" else "false";
```
Comment (SuperSandro2000) [inline nixos/modules/services/web-apps/mastodon.nix:35]: did you try to use ``toString (!cfg.s3.bucket)``

File: nixos/modules/services/web-apps/mastodon.nix [hunk 4]
```diff
@@ -20,20 +20,26 @@ let
...
+    S3_BUCKET= cfg.s3.bucket;
+    S3_HOSTNAME= cfg.s3.hostname;
+    S3_PROTOCOL= cfg.s3.protocol;
+    S3_REGION= cfg.s3.region;
+    SMTP_FROM_ADDRESS = cfg.smtp.fromAddress;
+    SMTP_PORT = toString(cfg.smtp.port);
```
Comment (SuperSandro2000) [inline nixos/modules/services/web-apps/mastodon.nix:41]:
```suggestion
    SMTP_PORT = toString cfg.smtp.port;
```

File: nixos/modules/services/web-apps/mastodon.nix [hunk 5]
```diff
@@ -403,6 +409,44 @@ in {
         };
       };
 
+      aws = {
+        SecretAccessKeyFile = lib.mkOption {
+          description = lib.mdDoc ''
+            Path to file containing the SecretAccessKeyFile.
+          '';
+          default = "/var/lib/mastodon/secrets/mastodon-secret-access-key";
+          example = "/run/keys/mastodon-smtp-password";
```
Comment (georgyo) [inline nixos/modules/services/web-apps/mastodon.nix:418]: There isn't and can't be an option to auto-generate this file, so having a default is not useful. The example is also still smtp.

I would delete the default and update the example.

**Comment by AndrewKvalheim** [comment]
What happened here? Is there still interest in supporting this?

**Comment by georgyo** [comment]
Having variables for this would be nice, but it is also easy enough to put these options in extra conf section.

This is how I have this deployed:

 https://github.com/georgyo/nix-conf/blob/c96147c8cfdd31fbf0c1640d4e91e3724fe34d86/hosts/hydra/mastodon/default.nix#L31

**Pull request closed without merging**